crates/types/src/cluster_controller.rs

```diff
@@ -205,3 +207,123 @@ impl From<SchedulingPlan> for SchedulingPlanBuilder {
         }
     }
 }
+
+/// Nodeset selection strategy for picking cluster members to host replicated logs. Note that this
+/// concerns loglet replication configuration across storage servers during log bootstrap or cluster
+/// reconfiguration, for example when expanding capacity.
+///
+/// It is expected that the Bifrost data plane will deal with short-term server unavailability.
+/// Therefore, we can afford to aim high with our nodeset selections and optimise for maximum
+/// possible fault tolerance. It is the data plane's responsibility to achieve availability within
+/// this nodeset during periods of individual node downtime.
+///
+/// Finally, nodeset selection is orthogonal to log sequencer placement.
+#[derive(Debug, Clone, Copy, Default, serde::Serialize, serde::Deserialize)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[serde(rename_all = "kebab-case")]
+pub enum NodeSetSelectionStrategy {
+    /// Selects an optimal nodeset size based on the replication factor. The nodeset size is at
+    /// least 2f+1, calculated by working backwards from a replication factor of f+1. If there are
+    /// more nodes available in the cluster, the strategy will use them.
+    ///
+    /// This strategy will never suggest a nodeset smaller than 2f+1, thus ensuring that there is
+    /// always plenty of fault tolerance built into the loglet. This is a safe default choice.
+    #[default]
+    StrictFaultTolerantGreedy,
+}
+
+#[serde_as]
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[cfg_attr(
+    feature = "schemars",
+    schemars(rename = "ClusterConfiguration", default)
+)]
+#[serde(rename_all = "kebab-case")]
+pub struct ClusterConfigurationSeed {
+    /// # Partitions
+    ///
+    /// Number of partitions that will be provisioned during cluster bootstrap,
+    /// partitions used to process messages.
+    ///
+    /// NOTE: This config entry only impacts the initial number of partitions.
+    /// Repartitioning is not supported yet
+    ///
+    /// Cannot be higher than `65535` (You should almost never need as many partitions anyway)
+    pub num_partitions: NonZeroU16,
+
+    /// # Default Provider
+    ///
+    /// The default bifrost provider
+    ///
+    /// Default: [`ProviderKind::Replicated`]
+    pub default_provider: ProviderKind,
```

Comment:
How do these settings relate to the existing `bootstrap_num_partitions` and the `default_provider` in the bifrost configuration?

Reply:
tbh, i am not sure yet how to handle the configuration migration fully yet. Those params definitely belong to the cluster configuration, so I assume the right thing to do is to use them as default values for the default cluster configuration. Unless overridden explicitly by the user.

Reply:
Right now these are respected when `cluster-configuration` is set to `default`.